src/tree_cmd.cpp

```diff
@@ -726,86 +782,126 @@ static void TileLoop_Trees(TileIndex tile)
+						case 2: { // add a neighbouring tree
+							spread = can_plant_extra;
+							break;
+						}
 
-						break;
+						default:
+							return;
 					}
-
-					default:
-						return;
 				}
-			}
-			break;
+				break;
 
-		case 6: // final stage of tree destruction
-			if (!CanPlantExtraTrees(tile)) {
-				/* if trees can't spread just plant a new one to prevent deforestation */
-				SetTreeGrowth(tile, 0);
-			} else if (GetTreeCount(tile) > 1) {
-				/* more than one tree, delete it */
-				AddTreeCount(tile, -1);
-				SetTreeGrowth(tile, 3);
-			} else {
-				/* just one tree, change type into MP_CLEAR */
-				switch (GetTreeGround(tile)) {
-					case TREE_GROUND_SHORE: MakeShore(tile); break;
-					case TREE_GROUND_GRASS: MakeClear(tile, CLEAR_GRASS, GetTreeDensity(tile)); break;
-					case TREE_GROUND_ROUGH: MakeClear(tile, CLEAR_ROUGH, 3); break;
-					case TREE_GROUND_ROUGH_SNOW: {
-						uint density = GetTreeDensity(tile);
-						MakeClear(tile, CLEAR_ROUGH, 3);
-						MakeSnow(tile, density);
-						break;
-					}
-					default: // snow or desert
-						if (_settings_game.game_creation.landscape == LT_TROPIC) {
-							MakeClear(tile, CLEAR_DESERT, GetTreeDensity(tile));
-						} else {
-							uint density = GetTreeDensity(tile);
-							MakeClear(tile, CLEAR_GRASS, 3);
-							MakeSnow(tile, density);
-						}
-						break;
+			case 6: // fina)l stage of tree destruction
+				if (!can_plant_extra) {
+					/* if trees can't spread just plant a new one to prevent deforestation */
+					SetTreeGrowth(tile, 0);
+				} else if (count > 1) {
+					/* more than one tree, delete it */
+					AddTreeCount(tile, -1);
+					SetTreeGrowth(tile, 3);
+				} else {
+					/* no more trees, mark the tile for clearing */
+					clear = true;
 				}
+				break;
+
+			default:
+				AddTreeGrowth(tile, 1);
+				break;
+		}
+	}
+
+	if (clear) {
+		/* All trees died, change type into MP_CLEAR */
+		switch (GetTreeGround(tile)) {
+			case TREE_GROUND_SHORE: MakeShore(tile); break;
+			case TREE_GROUND_GRASS: MakeClear(tile, CLEAR_GRASS, GetTreeDensity(tile)); break;
+			case TREE_GROUND_ROUGH: MakeClear(tile, CLEAR_ROUGH, 3); break;
+			case TREE_GROUND_ROUGH_SNOW: {
+				uint density = GetTreeDensity(tile);
+				MakeClear(tile, CLEAR_ROUGH, 3);
+				MakeSnow(tile, density);
+				break;
 			}
-			break;
+			default: // snow or desert
+				if (_settings_game.game_creation.landscape == LT_TROPIC) {
+					MakeClear(tile, CLEAR_DESERT, GetTreeDensity(tile));
+				} else {
+					uint density = GetTreeDensity(tile);
+					MakeClear(tile, CLEAR_GRASS, 3);
+					MakeSnow(tile, density);
+				}
+				break;
+		}
+	}
 
-		default:
-			AddTreeGrowth(tile, 1);
-			break;
+	if (spread) {
+		TreeType treetype = GetProceduralTreeType(tile, tile_hash);
+
+		/* We never spread while changing the current tile so it's ok to overwrite and only mark the new tile dirty later. */
+		tile += TileOffsByDir((Direction)(Random() & 7));
```

Comment:
Should this 7 be TREE_GROWTH_PROCEDURAL?

Reply:
No, that's direction, Again, not a new code. Though I guess this is easy to improve a bit.